pkg/services/annotations/annotationsimpl/loki/historian_store.go

```diff
@@ -48,13 +70,215 @@ func NewLokiHistorianStore(cfg setting.UnifiedAlertingStateHistorySettings, ft f
+		if err != nil {
+			if errors.Is(err, ErrMissingRule) {
+				return make([]*annotations.ItemDTO, 0), ErrLokiStoreInvalidQuery.Errorf("rule with ID %d does not exist", query.AlertID)
+			}
+			return make([]*annotations.ItemDTO, 0), ErrLokiStoreInternal.Errorf("failed to query rule: %w", err)
+		}
+	}
+
+	logQL, err := historian.BuildLogQuery(buildHistoryQuery(ctx, query, accessResources.Dashboards, rule.UID))
+	if err != nil {
+		return make([]*annotations.ItemDTO, 0), ErrLokiStoreInternal.Errorf("failed to build loki query: %w", err)
+	}
+
+	now := time.Now().UTC()
+	if query.To == 0 {
+		query.To = now.UnixMilli()
+	}
+	if query.From == 0 {
+		query.From = now.Add(-defaultQueryRange).UnixMilli()
+	}
+
+	// query.From and query.To are always in milliseconds, convert them to nanoseconds for loki
+	from := query.From * 1e6
+	to := query.To * 1e6
+
+	res, err := r.client.RangeQuery(ctx, logQL, from, to, query.Limit)
+	if err != nil {
+		return make([]*annotations.ItemDTO, 0), ErrLokiStoreInternal.Errorf("failed to query loki: %w", err)
+	}
+
+	items := make([]*annotations.ItemDTO, 0)
+	for _, stream := range res.Data.Result {
+		items = append(items, r.annotationsFromStream(ctx, stream, query.OrgID, *accessResources)...)
+	}
+
+	// order by time desc
+	sort.Slice(items, func(i, j int) bool {
+		return items[i].Time > items[j].Time
+	})
+
+	return items, err
+}
+
+func (r *LokiHistorianStore) annotationsFromStream(ctx context.Context, stream historian.Stream, orgID int64, ac accesscontrol.AccessResources) []*annotations.ItemDTO {
+	items := make([]*annotations.ItemDTO, 0, len(stream.Values))
+	for _, sample := range stream.Values {
+		entry := historian.LokiEntry{}
+		err := json.Unmarshal([]byte(sample.V), &entry)
+		if err != nil {
+			// bad data, skip
+			continue
+		}
+
+		if !hasAccess(entry, ac) {
+			// no access to this annotation, skip
+			continue
+		}
+
+		transition, err := buildTransition(entry)
+		if err != nil {
+			// bad data, skip
+			continue
+		}
+
+		if !historian.ShouldRecordAnnotation(*transition) {
+			// skip non-annotation transition
+			continue
+		}
+
+		annotationText, annotationData := historian.BuildAnnotationTextAndData(
+			historymodel.RuleMeta{
+				Title: entry.RuleTitle,
+			},
+			transition.State,
+		)
+
+		items = append(items, &annotations.ItemDTO{
+			AlertID:      entry.RuleID,
+			DashboardID:  ac.Dashboards[entry.DashboardUID],
+			DashboardUID: &entry.DashboardUID,
+			PanelID:      entry.PanelID,
+			NewState:     entry.Current,
+			PrevState:    entry.Previous,
+			Time:         sample.T.UnixMilli(),
+			Text:         annotationText,
+			Data:         annotationData,
+		})
+	}
+
+	return items
 }
 
 func (r *LokiHistorianStore) GetTags(ctx context.Context, query *annotations.TagsQuery) (annotations.FindTagsResult, error) {
 	return annotations.FindTagsResult{}, nil
 }
 
+// util
+
+func getRule(ctx context.Context, sql db.DB, orgID int64, ruleID int64) (*ngmodels.AlertRule, error) {
```

Comment:
It looks like we only need the uid from the rule. Maybe save some flops just have the function query for the uid of the provided id, WDYT?